### PR TITLE
Fix some quadratic performance bugs

### PR DESCRIPTION
**Pull request by gergo-**
When generating large functions, I noticed that Csmith can be excessively slow. For example, `csmith --max-funcs 1 --nomain --max-block-size 20 --seed 827638045` takes 4 minutes on my machine, and `csmith --max-funcs 1 --nomain --max-block-size 15 --seed 1229442881` even 7 minutes.

I suspected the use of quadratic algorithms, and profiling confirmed that in many places vectors were used to represent sets and searched linearly many times.

This series of patches fixes the worst of these cases by replacing vectors with sets. This brings both of the examples down to 1 minute each. (Which still seems a lot for generating 4000 lines of code.)

### CODE REVIEW

**Comment by jxyang** [comment]
Thanks for the improvements. I agree set is in general a better data structure than vector in these cases. I will review the changes soon.

Csmith does back tracking when some code generated later violated eariler assumptions. I am pretty sure  in the cases you discovered, Csmith had to back track by replacing a large chunk of code with a simple variable. That's why it appears taking unreasonable amount of time generating a few thousands lines of code.

**Comment by jxyang** [comment]
I will review the change over this weekend.

File: src/Block.cpp [hunk 1]
```diff
@@ -238,6 +238,7 @@ Block::~Block(void)
 	deleted_stms.clear();
 
 	local_vars.clear();
+	local_var_set.clear();
```
Comment (jxyang) [inline src/Block.cpp]: I don't think we need to maintain two collections of local variables. Local variables as a set is preferred.

File: src/Block.cpp [hunk 2]
```diff
@@ -491,6 +492,19 @@ Block::append_nested_loop(CGContext& cg_context)
...
 
+/* Make sure the vector and the set contain the same elements, under the
+ * assumption that the vector can only ever grow. If the sizes of the
+ * containers are equal, this function assumes that they do contain the same
+ * elements. Otherwise, the elements of the vector are added to the set.
+ */
+static void
+ensure_vector_and_set_match(const vector<Variable *>& vec, set<Variable *>& set)
+{
+	if (vec.size() != set.size()) {
```
Comment (jxyang) [inline src/Block.cpp]: This becomes unnecessary if we don't maintain both a vector and a set for local variables.

File: src/FactMgr.cpp [hunk 3]
```diff
@@ -513,15 +513,23 @@ void
 FactMgr::makeup_new_var_facts(vector<const Fact*>& old_facts, const vector<const Fact*>& new_facts)
 {
     size_t i;
+    // We will search the old facts repeatedly. Copy the entries to a map
+    // for faster lookup.
+    FactMap old_fact_map;
```
Comment (jxyang) [inline src/FactMgr.cpp]: I believe, instead of creating a map on demand,  we can get much better speed up by storing facts in a map at the first place. This requires replacing all 'FactVec' with 'FactMap'.

File: src/Variable.cpp [hunk 4]
```diff
@@ -616,43 +647,39 @@ vector<string>
 Variable::deputy_annotation(void) const
```
Comment (jxyang) [inline src/Variable.cpp]: I don't feel the need to support Deputy in Csmith anymore. I will remove it soon.

Comment (regehr) [inline src/Variable.cpp]: Yes!

File: src/Block.cpp (hunk 1 above)
Comment (gergo-) [inline src/Block.cpp]: I agree. There was some reason that I didn't rip out the old collections completely; I think it was simply a *lot* of work in a lot of places. In these patches I only did the parts that seemed to make a difference in performance according to my profiling.
Thanks for the review, I will see when I get around to reworking the issues you mentioned.

Comment (jxyang) [inline src/Block.cpp]: I know it's a much larger scope than you originally proposed. The original data structure, vector, was chosen without consideration of performance. Thanks for making this right!

**Comment by jxyang** [comment]
Csmith has performance problems for sure. I will pick up this PR when I have time.

**Comment by jxyang** [comment]
Finally I get the desired changes implemented - basically replacing `std::vector<Variable*>` with `std::unordered_set<Variable*>`. The changes are in branch `jxyang.variableSet`, involving thousands of lines. However, the experimental results are disappointing: generation of 1000 test cases of same set of seeds takes slightly longer with the changes. 43 vs 40 minutes to be exact.

My hypothesis is that even though the changes improved some edge cases, it slows down the common cases. Therefore, the overall performance benefit is gone.

I'd love to hear some suggestions on further performance improvement (or whether I implemented the wrong way).

**Comment by jxyang** [comment]
https://github.com/csmith-project/csmith/tree/jxyang.variableSet

**Comment by jxyang** [comment]
Closing for now due to undecisive experimental results.
